sdk/identity/src/token_credentials/auto_refreshing_credentials.rs

```diff
@@ -38,38 +40,139 @@ impl AutoRefreshingTokenCredential {
 #[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
 impl TokenCredential for AutoRefreshingTokenCredential {
     async fn get_token(&self, resource: &str) -> azure_core::Result<TokenResponse> {
-        // if the current cached token is good, return that.
-        if let Some(Ok(token)) = self.current_token.read().await.as_ref() {
+        // if the current cached token for this resource is good, return it.
+        let token_cache = self.token_cache.read().await;
+        if let Some(token) = token_cache.get(resource) {
             if !is_expired(token) {
                 return Ok(token.clone());
             }
         }
 
-        let mut guard = self.current_token.write().await;
+        // otherwise, drop the read lock and get a write lock to refresh the token
+        drop(token_cache);
+        let mut token_cache = self.token_cache.write().await;
 
         // check again in case another thread refreshed the token while we were
         // waiting on the write lock
-        if let Some(Ok(token)) = guard.as_ref() {
+        if let Some(token) = token_cache.get(resource) {
             if !is_expired(token) {
                 return Ok(token.clone());
             }
         }
 
-        let res = self.credential.get_token(resource).await;
+        let result = self.credential.get_token(resource).await;
```

Comment:
What do you think about bubbling out the error using `?`?  The match behavior was important before, when this cached the `Result`, but I think this might be more readable.


```suggestion
        let token = self.credential.get_token(resource).await?;
         // NOTE: we do not check to see if the token is expired here, as at
         // least one credential, `AzureCliCredential`, specifies the token is
         // immediately expired after it is returned, which indicates the token
         // should always be refreshed upon use.        
        token_cache.insert(resource.to_string(), token.clone());
        Ok(token)
```

Note: the GitHub UI makes making a multi-line suggestion that includes new, unchanged, and removed code behave weirdly, so this patch suggestion does not remove the match code later on.

Reply:
The difference here is that this code bubbles out the underlying token error instead of wrapping in another error type.

Reply:
Yep - I agree.  I have updated.

I'm also tempted to remove the `NOTE:...` comment, as I think it is confusing (why would you check a newly created token for expiry?) and possibly out of date.  I've just tested `az account get-access-token` in Linux and Windows and both return an access token with a 1 hour expiry.  Thoughts?

Reply:
Hmm... I see that the `NOTE:` comment was added as part of the fix for this issue:
    https://github.com/Azure/azure-sdk-for-rust/issues/1143

I remember seeing this issue, and thinking that it must be related to the switch from `chrono` to `time`, and handling of timezones when parsing the returned expiry time.  I did look at the code and couldn't see anything obviously wrong with it, but might take another look.

Reply:
LGTM